### PR TITLE
grid_sample: support pixel-coordinate grids

### LINKED ISSUE #36107: Allow grid_sample to accept pixel units (absolute coordinates)
_This issue is expanded from #24870 for reference and for additional discussions on implementation details. See that issue for context._

Add an option for the grid/flow input to `grid_sample` to be expressed in pixel units (absolute coordinates). This is much more natural for some applications, including image registration, where the grid-like input is actually a displacement field, and this eliminates the need for the user to convert to normalized [-1,1] coordinates, which is prone to user error, (especially when users don’t know which setting of `align_corners` to target).

This would help simplify the implementation of convolutional networks to do Optical Flow Estimation and Image Registration tasks.

### PR DESCRIPTION
**Pull request by vboussot**
## Issue

Fixes #36107.

Pixel-coordinate sampling was first requested in #4085 (2017) and accepted in the 2019 grid_sample triage, #24870: "Yes absolute coordinate is great, we want this." A previous implementation, #41241, was approved and then stalled on the BC checks tripped by widening the existing operator schemas; this PR introduces private ops instead of touching those schemas.

Also addresses #81868 and #127031. Related: #61528, #94733, #149013. The relative/displacement grid of #94816 and #36108 (accepted, never implemented) becomes a one-flag follow-up on these ops: a displacement in pixel units is the identity plus the offset, added inside the kernel in opmath, which is what makes a half-precision flow field usable.

## Summary

This PR adds an opt-in pixel-coordinate route to `F.grid_sample`:

```python
out = F.grid_sample(input, grid, mode="bicubic", pixel_coords=True, cubic_coeff_a=-0.5)
```

With `pixel_coords=True` the grid holds continuous pixel indices of the input instead of `[-1, 1]` coordinates: `x = 0` is the center of the first pixel and `x = W - 1` the center of the last, at any resolution. Two properties follow that normalized coordinates cannot offer: sampling at integer locations reproduces the input bitwise in float32 (#81868), and cropping a window that holds every tap while shifting the grid by the offset yields bitwise-identical samples, which is what tiled and out-of-core pipelines need (one exception: `nearest` rounds a coordinate exactly halfway between two pixels to the even one, as the normalized route does, so such a tie follows the parity of the offset). `align_corners` keeps only its reflection-padding role; all padding modes behave exactly as on the normalized route.

The new route also accepts a `torch.double` grid over any supported floating payload (float64/32/16, bfloat16), keeping coordinate precision independent of the payload dtype on large volumes, and `cubic_coeff_a` selects the Keys coefficient of the bicubic kernel (`-0.5` is Catmull-Rom), so a resample can match external cubic implementations instead of being locked to `-0.75`. Under autocast, input and grid are promoted to the wider of their two dtypes, except that a double grid stays double and leaves the input as it is.

As a concrete consumer, [KonfAI](https://github.com/fideus-labs/KonfAI) (medical imaging, out-of-core inference) hand-rolls exactly this sampler today, a 64-tap Catmull-Rom gather in global float64 pixel coordinates, and can collapse it onto the native kernel.

Implemented as four private CPU/CUDA ops (`_grid_sampler_{2d,3d}_pixel` and backwards) with double backward. `pixel_coords` and the grid dtype are template parameters of the kernels on both devices, so the normalized instantiations contain no pixel branch and keep the fixed-coefficient cubic helpers; their CUDA register counts are unchanged and the normalized route is bit-identical.

Per the FC policy the `native_functions.yaml` notice cites, the functional (`torch/nn/functional.py`, `functional.pyi.in`, `overrides.py` and the surface test) is one commit and can be split into a second PR to land two weeks after the operators, if preferred.

## Testing

- Parity against the normalized route on the same coordinates: forward and first-order gradients, 2-D/3-D x {bilinear, nearest, bicubic} x {zeros, border, reflection} x `align_corners` x in- and out-of-range, CPU and CUDA: 72 forward and 48 first-order gradient cases, at 6e-15 and 2e-14 (double).
- Device-type tests: `test_grid_sample_pixel_parity` (vs the normalized route on the same coordinates, out of range included, per mode/padding/align_corners), `test_grid_sample_pixel_exactness` (bitwise integer locations including a non-binary `cubic_coeff_a=-0.3`, 2-D and 3-D, and the double backward's exact zeros on the zero-weight taps of an integer location), `test_grid_sample_pixel_keys_coefficient` (the kernel against a straight-line Keys reference), `test_grid_sample_pixel_tiling_invariance` (crop/shift bitwise, up to volumetric bicubic), `test_grid_sample_pixel_grad` (gradcheck + gradgradcheck per dim/mode/padding with straddling grids), `test_grid_sample_pixel_grad_on_the_edge` (a constant image has no grid gradient on the first and last pixel, border and reflection), `test_grid_sample_pixel_dtypes`, `test_grid_sample_pixel_mixed_double_backward`, `test_grid_sample_pixel_double_grid_precision` (a coordinate only float64 can resolve), `test_grid_sample_pixel_autocast`, `test_grid_sample_pixel_non_contiguous`, `test_grid_sample_pixel_non_finite` (NaN/Inf/huge coordinates forward and backward, in half and bfloat16 too, and the Inf-aliasing case through the composite double backward), `test_grid_sample_pixel_non_finite_double_backward` (a non-finite coordinate has the second order of the far coordinate the kernels sample, and one past the index type that of any coordinate out on its side), `test_grid_sample_pixel_bicubic_double_backward_low_precision`, `test_grid_sample_pixel_bicubic_last_voxel_of_a_wide_axis` (2**24 + 1 and 2**53 + 1); plus a surface test for rejections, meta shapes, scripting, and a fullgraph torch.compile smoke.
- The ops are not in the OpInfo db: they are CPU/CUDA-only and pixel samples in `sample_inputs_grid_sample` would drag mps/xpu xfails behind them. A dedicated OpInfo is a fine follow-up if preferred.

## Benchmark

`torch.utils.benchmark` medians on the same coordinates, `border`/`align_corners=False`, N=1 C=1 float32, RTX PRO 5000 Blackwell (laptop) + Core Ultra 9 275HX. Reported as pixel/normalized time ratios: ratios are stable across repeated order-swapped runs where absolute times vary with laptop thermals. For scale, 3-D bicubic forward is ~1.3-1.9 ms at 128³.

| same coordinates, pixel/norm ratio | fwd | fwd+bwd |
|---|---|---|
| CUDA 3-D 128³ bilinear | 1.00 | 0.91-1.00 |
| CUDA 3-D 128³ nearest | 0.97 | |
| CUDA 3-D 128³ bicubic | 0.86-0.98 | 0.99-1.00 |
| CUDA 2-D 2048² bilinear | 0.93 | 0.96 |
| CUDA 2-D 2048² bicubic | **0.62-0.65** | **0.61-0.71** |
| CPU 3-D 64³ bilinear / bicubic | 0.97 / 0.86 | |
| CPU 2-D 1024² bilinear / bicubic | 5.2 / 3.7 | |

The double-grid case (today: cast the whole payload to f64) at C=8, bicubic 3-D 128³: **pixel f32 + f64 grid is 2.5-2.8x faster** (e.g. 20.2 vs 50.9 ms) **at lower peak memory** (0.39 vs 0.60 GB) than the normalized all-f64 route. Known performance asymmetry: the 2-D CPU pixel route is the scalar implementation, 3.7-5.2x behind the hand-vectorized normalized 2-D path (vectorizing it is a follow-up; 3-D CPU, scalar vs scalar, has pixel slightly ahead).

## Limitations / follow-ups

vmap and Inductor fall back to the op. No forward AD yet (#195280 is adding it to the normalized route; the pixel ops can follow its rules). MPS/XPU are not implemented. On CUDA, bilinear and nearest keep the `int` bounds of the merged kernels, so an axis past INT_MAX samples as out of range there, as it does on the normalized route; bicubic indexes in `index_t`. The double backward over a double grid casts the payload to double, so the memory gain of the mixed route is first order only. The C++ `torch::nn::functional::grid_sample` options are unchanged, and ONNX export of the pixel route is not wired (the exporter's symbolic covers `aten::grid_sampler` only). The relative-grid flag (#94816) and a vectorized 2-D CPU path are natural follow-ups.

## Checklist

- [x] Passes lint (`spin fixlint`)
- [x] Added/updated tests
- [x] Updated documentation (if applicable)
- [x] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No. The default path is unchanged down to the bit; the new keywords are opt-in, and `cubic_coeff_a` raises anywhere it would not be honored.

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

**Comment by kurtamohler** [comment]
You can use [ghstack](https://github.com/ezyang/ghstack) to stack PRs so that the diff only shows the changes intended for this PR

**Comment by vboussot** [comment]
Hi @kurtamohler, ghstack needs write access to the repo it opens the PRs in, which I don't have, so the stacked branches it would create in pytorch/pytorch aren't an option here.

It is stacked all the same, on #194787, so this drops to its own 6 commits once that lands and I rebase. Until then: https://github.com/vboussot/pytorch/compare/gridsample-tricubic...gridsample-pixel-coords
